### ISSUE TITLE
noqa syntax-check does not work for syntax-check

### ISSUE THREAD

**Issue by rokcarl** [ansible-community/ansible-lint#1506]
##### Summary

Running `ansible-lint -p playbooks/app/info.yml` I get:

```
roles/app/tasks/info.yml:31:3: syntax-check The field 'hosts' has an invalid value, which includes an undefined variable. The error was: 'hostvars' is undefined
```

The underlying code is:

```
# noqa syntax-check
- name: Gather host info # noqa syntax-check
  hosts: "{{ hostvars['127.0.0.1']['server_master_name'] }},{{ hostvars['127.0.0.1']['server_standby_name'] }}" # noqa syntax-check
  vars_files:
    - ../../../global_vars/app.yml
  tasks:
    ...
```

The line with the `- name: ...` is line from the error.

So I put ` # noqa syntax-check` on line 31 at the end, it doesn't solve it, so I put it at a line before and a line after as well, doesn't work. I see that the number for syntax check [is 911 in the code](https://github.com/ansible-community/ansible-lint/blob/878a90edbce2a985837d26883bff636ec9c7c3a7/src/ansiblelint/constants.py#L120), so I try ` # noqa 911`, doesn't work either.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
$ ansible --version
ansible 2.10.7
  config file = /ansible/ansible.cfg
  configured module search path = ['/home/ci-user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.9.2 (default, Mar 31 2021, 12:21:33) [GCC 8.3.0]

$ ansible --version
ansible 2.10.7
  config file = /ansible/ansible.cfg
  configured module search path = ['/home/ci-user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.9.2 (default, Mar 31 2021, 12:21:33) [GCC 8.3.0]
ci-user@af7ace40ea90:/ansible$ ansible-lint --version
ansible-lint 5.0.6 using ansible 2.10.7

```

- ansible installation method: pip
- ansible-lint installation method: pip

##### OS / ENVIRONMENT

This is inside a docker container from the `python:3.9.2-slim-buster` image, which is built upon `debian:buster-slim`.

##### STEPS TO REPRODUCE

```bash
$ cat a.yml
# noqa syntax-check
- name: Gather host info # noqa syntax-check
  hosts: "{{ hostvars['127.0.0.1']['server_master_name'] }},{{ hostvars['127.0.0.1']['server_standby_name'] }}" # noqa syntax-check
  tasks:

    - name: Is app running? # noqa 305 306
      shell: curl localhost:9000 -sI | head -n 1 | awk '{print $2}'
      register: http_status_raw
      args:
        warn: false

$ ansible-lint -p a.yml
WARNING  Overriding detected file kind 'yaml' with 'playbook' for given positional argument: a.yml
WARNING  Listing 1 violation(s) that are fatal
a.yml:2:3: syntax-check The field 'hosts' has an invalid value, which includes an undefined variable. The error was: 'hostvars' is undefined
You can skip specific rules or tags by adding them to your configuration file:
# .ansible-lint
warn_list:  # or 'skip_list' to silence them completely
  - syntax-check  # Ansible syntax check failed
Finished with 1 failure(s), 0 warning(s) on 1 files.
```

**Comment by rokcarl** [comment]
I even added it to the skip list in `.ansible-lint`, but it still complains:
```
skip_list:
  - line-length
  - command-instead-of-shellconsistent
  - syntax-check
```

**Comment by bendem** [comment]
This is problematic because syntax-check is fully broken if you have vaulted files in a `vars_files` block. In our case, about half of our playbooks can't be linted because of this.

**Comment by bendem** [comment]
Also, `syntax-check` failing hides all other errors as can be seen below:

```shell
(ansible) ➜  ansible git:(main) ✗ ansible-lint -c .ansible-lint roles/*                                           
Loading custom .yamllint.yml config file, this extends our internal yamllint config.
WARNING  Listing 1 violation(s) that are fatal
roles/springboot-app/molecule/default/converge.yml:0: internal-error Unexpected error code 1 from execution of: ansible-playbook --syntax-check roles/springboot-app/molecule/default/converge.yml
You can skip specific rules or tags by adding them to your configuration file:
# .ansible-lint
warn_list:  # or 'skip_list' to silence them completely
Finished with 0 failure(s), 1 warning(s) on 217 files.

(ansible) ➜  ansible git:(main) ✗ ansible-lint -c .ansible-lint roles/^(springboot-app)*
Loading custom .yamllint.yml config file, this extends our internal yamllint config.
WARNING  Listing 13 violation(s) that are fatal
playbooks/global.yml:32: unnamed-task All tasks should be named
playbooks/global.yml:36: unnamed-task All tasks should be named
roles/httpd-vhosts/meta/main.yml:0: role-name Role name httpd-vhosts does not match ``^+$`` pattern
roles/java/molecule/default/verify.yml:12: no-changed-when Commands should not change things if nothing needs doing
roles/java/molecule/default/verify.yml:20: risky-file-permissions File permissions unset or incorrect
roles/java/molecule/default/verify.yml:20: unnamed-task All tasks should be named
roles/java/molecule/default/verify.yml:36: no-changed-when Commands should not change things if nothing needs doing
roles/keepalived/molecule/default/side_effect.yml:8: risky-file-permissions File permissions unset or incorrect
roles/ldap_user/molecule/default/converge.yml:16: unnamed-task All tasks should be named
roles/ldap_user/molecule/default/verify.yml:11: unnamed-task All tasks should be named
roles/logrotate/molecule/default/verify.yml:6: no-changed-when Commands should not change things if nothing needs doing
roles/sslcerts/tasks/03-renew_script.yml:49: no-handler Tasks that run when changed should likely be handlers
roles/talend-runner:0: role-name Role name talend-runner does not match ``^+$`` pattern
You can skip specific rules or tags by adding them to your configuration file:
# .ansible-lint
warn_list:  # or 'skip_list' to silence them completely
  - experimental  # all rules tagged as experimental
  - no-changed-when  # Commands should not change things if nothing needs doing
  - no-handler  # Tasks that run when changed should likely be handlers
  - risky-file-permissions  # File permissions unset or incorrect
  - role-name  # Role name {0} does not match ``^[a-z][a-z0-9_]+$`` pattern
  - unnamed-task  # All tasks should be named
Finished with 13 failure(s), 0 warning(s) on 207 files.
```

**Comment by phospi** [comment]
@rokcarl I agree that skipping `syntax-check` seems broken. However, I feel the need to address your example for all people browsing this issue. `syntax-check` is very picky about variable definition. When a variable is undefined, there is no need to disable `syntax-check`. The configuration file `.ansible-lint` has a dedicated parameter to define variables. 

Please consider this and think twice before disabling `syntax-check`.

https://ansible-lint.readthedocs.io/en/latest/configuring.html
````
# Define required Ansible's variables to satisfy syntax check
extra_vars:
  foo: bar
  multiline_string_variable: |
    line1
    line2
  complex_variable: ":{;\t$()"
````

I'm not sure about the yaml syntax, but a variable definition in your case should look similar to this:
````
# Define required Ansible's variables to satisfy syntax check
extra_vars:
  hostvars:
    127.0.0.1:
       server_master_name: my_server_master_name

````

**Comment by rokcarl** [comment]
Not a viable solution for me, this would define the variables statically in the config file, but the variables are dynamic and already set at the point where Ansible complains.

Either way, wouldn't want to put the variable definition in the config file.

**Comment by bendem** [comment]
Indeed, a good solution for us would be for syntax check to work correctly. Disabling it is last resort, but it should work.

The check can already ignore includes it can't resolve, it could just as well ignore include that were vaulted.

This fails:
```yml
vars_files:
    - "./vault/something.vault"
```
This doesn't

```yml
vars_files:
    - "./vault/something-{{ env }}.vault"
```

**Comment by ssbarnea** [comment]
syntax-check cannot be disabled using noqa because it is produces by ansible itself, before even we use our own loader. The only way to avoid it is to prevent ansible from loading these files, excluding them or using kinds to help linter identify their type.

I am closing it as there is no action expected here, but this should not prevent anyone to make PR that refer this ticket.

**Comment by bendem** [comment]
> syntax-check cannot be disabled using noqa

What about disabling it altogether? It is broken in many cases and it prevents users from using `ansible-lint` at all.
